plugins/modem-manager/fu-firehose-updater.c

```diff
@@ -0,0 +1,699 @@
+				     program_filename);
+			return FALSE;
+		}
+
+		while ((payload_size + (guint)program_sector_size) < max_payload_size)
+			payload_size += (guint)program_sector_size;
+
+		g_debug ("sending program file '%s' (%zu bytes)", program_filename, g_bytes_get_size (program_file));
+		if (!fu_firehose_updater_send_program_file (self, program_filename, program_file,
+							    payload_size, program_sector_size, error)) {
+			g_prefix_error (error, "Failed to send program file '%s': ",
+					program_filename);
+			return FALSE;
+		}
+
+		g_clear_object (&rsp_node);
+		g_clear_object (&rsp_silo);
+
+		g_debug ("waiting for program file download confirmation...");
+		if (!fu_firehose_updater_send_and_receive (self, NULL, &rsp_silo, &rsp_node, error)) {
+			g_prefix_error (error, "Download confirmation not received for file '%s': ", program_filename);
+			return FALSE;
+		}
+
+		if (!fu_firehose_updater_check_operation_result (rsp_node, &rawmode)) {
+			g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+				     "Download confirmation failed for file '%s'",
+				     program_filename);
+			return FALSE;
+		}
+
+		if (rawmode != FALSE) {
+			g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+				     "Download confirmation failed for file '%s': rawmode still enabled",
+				     program_filename);
+			return FALSE;
+		}
+
+		sent_bytes += g_bytes_get_size (program_file);
+		g_signal_emit (self, signals [SIGNAL_WRITE_PROGRESS], 0,
+			       (guint)((100.0 * (gdouble)sent_bytes) / (gdouble)total_bytes));
+	}
+
+	return TRUE;
+}
+
+gboolean
+fu_firehose_updater_write (FuFirehoseUpdater *self, XbSilo *silo, GPtrArray *action_nodes, GError **error)
+{
+	guint max_payload_size;
+	gboolean result;
+	g_autoptr(GError) reset_error = NULL;
+
+	g_signal_emit (self, signals [SIGNAL_WRITE_PROGRESS], 0, 0);
+
+	if (!fu_firehose_updater_initialize (self, error))
+		return FALSE;
+
+	max_payload_size = fu_firehose_updater_configure (self, error);
+	if (max_payload_size == 0)
+		return FALSE;
+
+	result = fu_firehose_updater_run_actions (self, silo, action_nodes, max_payload_size, error);
+
+	if (!fu_firehose_updater_reset (self, &reset_error)) {
+		if (result)
+			g_propagate_error (error, g_steal_pointer (&reset_error));
+		return FALSE;
+	}
+
+	g_signal_emit (self, signals [SIGNAL_WRITE_PROGRESS], 0, 100);
+
+	return result;
+}
+
+static void
+fu_firehose_updater_init (FuFirehoseUpdater *self)
+{
+}
+
+static void
+fu_firehose_updater_finalize (GObject *object)
+{
+	FuFirehoseUpdater *self = FU_FIREHOSE_UPDATER (object);
+	g_warn_if_fail (self->io_channel == NULL);
+	g_free (self->port);
+	G_OBJECT_CLASS (fu_firehose_updater_parent_class)->finalize (object);
+}
+
+static void
+fu_firehose_updater_class_init (FuFirehoseUpdaterClass *klass)
+{
+	GObjectClass *object_class = G_OBJECT_CLASS (klass);
+	object_class->finalize = fu_firehose_updater_finalize;
+
+	signals [SIGNAL_WRITE_PROGRESS] =
+		g_signal_new ("write-progress",
```

Comment:
i think everywhere else it's called `percentage` no?

Reply:
thanks, fixed